<a href="https://colab.research.google.com/github/mdheying/356-Python/blob/main/Regression_with_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction
At this point in your career, you are likely comfortable with straight-line fitting (linear regression) via a spreadsheet program like Excel, etc.  In this course, we are not only interested in a particular value, but how well we know that value.  Thus, the uncertainty of a slope and intercept are quantities that we want to be able to determine.

# Regression in python
As with error propagation and uncertainty of repeated measurements, python provides a route to determining both the slope and intercept values and their respective uncertainties.

In the example below, we will fit a few data points that are typed into our code.  In practice, it is often more convenient to import your data.  As with the other techniques, the tools that we need are in SciPy.  Here, we import SciPy in its entirety.

In [ ]:
from scipy import *

# Example data
x = [4, 7, 11, 13]
y = [8.3, 16.3, 26.3, 30.1]

regress = stats.linregress(x,y)

slope = regress.slope
intercept = regress.intercept

print(slope,intercept)

2.444102564102564 -1.1358974358974372


In the block above, we fit the data with the *linregress* tool and extracted the slope and intercept.  The happy result here is that the additional quantities that we desire have already been determined - they are in the *regress* object.

In [ ]:
slope_uncertainty = regress.stderr
intercept_uncertainty = regress.intercept_stderr
print(slope_uncertainty,intercept_uncertainty)

0.09143749369445123 0.8614071939069521


Remember the error convention for this course means that we need to report these values based on the uncertainty and matching decimal places.  For the values calculated above: <br>
slope = 2.444 &plusmn; 0.091 <br>
intercept = -1.14 &plusmn; 0.86 <br>
Notice that their uncertainties dictate reporting them to different decimal places and different numbers of significant figures.